zomes/deepkey/src/keyset_root/validate.rs

```diff
@@ -26,109 +26,199 @@ impl TryFrom<&Element> for KeysetRoot {
+    }
+    else {
+        Ok(ValidateCallbackResult::Valid)
+    }
+}
+
+fn _validate_create_authorization(create_header: &Create, proposed_keyset_root: &KeysetRoot) -> ExternResult<ValidateCallbackResult> {
+    // The author must be the FDA.
+    if *proposed_keyset_root.as_first_deepkey_agent_ref() != create_header.author {
+        return Error::FdaAuthor.into();
+    }
+    // The signature must be correct.
+    else if !proposed_keyset_root.verify_signature()? {
+        return Error::FdaSignature.into();
+    }
+    else {
+        Ok(ValidateCallbackResult::Valid)
+    }
+}
+
 #[hdk_extern]
 /// Create only.
 fn validate_create_entry_keyset_root(validate_data: ValidateData) -> ExternResult<ValidateCallbackResult> {
-    let element = validate_data.element;
-    let (header_hashed, entry) = element.into_inner();
-    let (signed_header, _) = header_hashed.into_inner();
-
-    // Header is a create as per the validate callback name.
-    let create_header = match signed_header.header() {
-        Header::Create(create_header) => create_header,
-        // Holochain sent the wrong header!
-        _ => unreachable!(),
+    let proposed_keyset_root = match KeysetRoot::try_from(&validate_data.element) {
+        Ok(keyset_root) => keyset_root,
+        Err(e) => return Ok(ValidateCallbackResult::Invalid(e.to_string())),
     };
 
-    // Header needs to be in the correct position in the chain.
-    if create_header.header_seq != KEYSET_ROOT_CHAIN_INDEX {
-        return Ok(ValidateCallbackResult::Invalid(Error::Position(create_header.header_seq, KEYSET_ROOT_CHAIN_INDEX).to_string()));
-    }
+    if let Header::Create(create_header) = validate_data.element.header().clone() {
+        match _validate_create_header(&create_header) {
+            Ok(ValidateCallbackResult::Valid) => {},
+            validate_callback_result => return validate_callback_result,
+        }
 
-    let keyset_root = match entry {
-        ElementEntry::Present(serialized_keyset_root) => {
-            match KeysetRoot::try_from(serialized_keyset_root) {
-                Ok(keyset_root) => keyset_root,
-                Err(e) => return Ok(ValidateCallbackResult::Invalid(Error::Wasm(e).to_string()))
-            }
-        },
-        _ => return Ok(ValidateCallbackResult::Invalid(Error::EntryMissing.to_string()))
-    };
+        match _validate_create_authorization(&create_header, &proposed_keyset_root) {
+            Ok(ValidateCallbackResult::Valid) => {},
+            validate_callback_result => return validate_callback_result,
+        }
 
-    if !keyset_root.verify_signature()? {
-        return Ok(ValidateCallbackResult::Invalid(Error::FdaSignature.to_string()));
+        Ok(ValidateCallbackResult::Valid)
     }
-
-    if !(*keyset_root.as_first_deepkey_agent_ref() == create_header.author) {
-        return Ok(ValidateCallbackResult::Invalid(Error::FdaAuthor.to_string()))
+    // Holochain sent the wrong header!
+    else {
+        unreachable!();
```

Comment:
I question if we should just return `Invalid` even though it is very unlikely someone breaks holochain that badly :sweat_smile: